### PR TITLE
Implementing PReLu

### PR DESCRIPTION
**Pull request by MythicArrow**
Implementing the Parametric ReLU from the well-known paper "Delving Deep into Rectifiers: Surpassing Human-Level Performance on ImageNet Classification", which was proposed in ICCV 2015. This function introduces a parameter named "a" that is learnable, and it allows the function to adapt during training, potentially improving model accuracy and convergence compared to standard ReLU or Leaky ReLU functions.
Denoted as: f(x) = x if x>= 0 | ax if x < 0
ArXiv link: https://arxiv.org/abs/1502.01852#

### CODE REVIEW

**Comment by jakevdp** [comment]
Hi - thanks for the contribution! It looks like this would be a good contribution to `jax.nn`, but there are a number of changes we'd have to make: mainly, the `a` value should be an explicit parameter of the function, otherwise we wouldn't be able to differentiate with respect to it. With this in mind, the function should not take `init` or `rng` or `num_parameters` as arguments. Also, the implementation should probably modeled after that of the existing `relu`. We would also need to add tests for the new function, in `tests/nn_test.py`. Is this something you'd like to work on?

**Comment by MythicArrow** [comment]
Yes of course sir, I would.

**Comment by MythicArrow** [comment]
So would you willing to give me some time to work on it? Currently, I am a bit busy.

**Comment by MythicArrow** [comment]
Would like to hear your feedback!

**Comment by MythicArrow** [comment]
I will also add tests later.

**Comment by MythicArrow** [comment]
Ok the pr combination has been done.

File: jax/_src/nn/functions.py [hunk 1]
```diff
@@ -1417,3 +1417,21 @@ def scaled_dot_general(
   )
 
   return out
+  @jax.jit
+  def prelu(x: ArrayLike, a: ArrayLike) -> Array:
+    """
+    Applies the PReLU activation function element-wise.
```
Comment (jakevdp) [inline jax/_src/nn/functions.py]: We should mention what PReLU means, and also add the formula for it (see the `relu` docstring for an example).

File: jax/_src/nn/functions.py [hunk 2]
```diff
@@ -1417,3 +1417,21 @@ def scaled_dot_general(
...
+
+    Args:
+        x (jnp.ndarray): Input tensor.
+        a (jnp.ndarray): Learnable parameter.
```
Comment (jakevdp) [inline jax/_src/nn/functions.py]: Please use two-space indentations. Also "learnable parameter" is not particularly descriptive – maybe say "slope parameter" or something similar?

Additionally: we should not list the type in parentheses after the argument name (see other docstrings in this file for examples of proper formatting).

File: jax/_src/nn/functions.py [hunk 3]
```diff
@@ -1417,3 +1417,21 @@ def scaled_dot_general(
...
+                             
+
+    Returns:
+        jnp.ndarray: Output tensor with the PReLU activation applied.
```
Comment (jakevdp) [inline jax/_src/nn/functions.py]: Remove the `jnp.ndarray:`

File: jax/_src/nn/functions.py [hunk 4]
```diff
@@ -1417,3 +1417,21 @@ def scaled_dot_general(
...
+  """
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:1500]: We should add an example or two here.

File: jax/_src/nn/functions.py [hunk 5]
```diff
@@ -1417,3 +1417,21 @@ def scaled_dot_general(
...
+
+    x = jnp.asarray(x)
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:1502]: Use `check_arraylike` before passing these to `asarray`

File: jax/_src/nn/functions.py [hunk 6]
```diff
@@ -1417,3 +1417,21 @@ def scaled_dot_general(
...
+    a = jnp.asarray(a)
+    return jnp.where(x >= 0, x, a * x)
```
Comment (jakevdp) [inline jax/_src/nn/functions.py]: I'm not sure that this will be sufficient under autodiff: for example, `relu` has a custom JVP rule and some notes about its gradient behavior in the docstring. We should probably do similar here, but I'm not entirely sure what the best form to use is. `a * relu(x)` may be sufficient, but in other places it seems to be implemented as `max(0, x) + a * min(0, x)`.

It would probably require constructing a few examples with gradients to see which form works in practice.

File: tests/nn_test.py [hunk 7]
```diff
@@ -469,7 +469,21 @@ def testRelu6Grad(self):
...
+      
+     y = nn.prelu(x, a) 
+     self.assertAllClose(y, expected,check_dtypes=False) 
+  
+     y_scalar = nn.prelu(x, 0.1) 
+     self.assertAllClose(y_scalar, expected,check_dtypes=False) 
+    
+  def testPreluGrad(x, a): 
+     return jnp.sum(nn.prelu(x, a)) 
+     check_grads(testprelugrad, (x, a), order=1)
```
Comment (jakevdp) [inline tests/nn_test.py]: This test looks strange – copy-paste problem?

Comment (MythicArrow) [reply tests/nn_test.py]: Yeah, after seeing the test file I realized that the tests for different purposes like grad and value were separated so I did the same.

File: jax/_src/nn/functions.py (hunk 5 above)
Comment (MythicArrow) [reply jax/_src/nn/functions.py:1502]: I had actually used it in the beginning of the function in the parenthesis.

File: jax/_src/nn/functions.py (hunk 6 above)
Comment (MythicArrow) [reply jax/_src/nn/functions.py]: The a*relu(x) may be incorrect because I checked the official paper and PyTorch implements it correctly. So should I do the same?

File: jax/_src/nn/functions.py (hunk 5 above)
Comment (MythicArrow) [reply jax/_src/nn/functions.py:1502]:
```
def prelu(x: ArrayLike, a: ArrayLike) -> Array:
```
Here I had used it.

File: jax/_src/nn/functions.py (hunk 6 above)
Comment (MythicArrow) [reply jax/_src/nn/functions.py]: By the way, I checked the derivative of prelu and it was 1 for positive inputs and for negative inputs it was a, the learned slope.
So should I do it according to its grad?

**Comment by MythicArrow** [comment]
Hello sir, have you been busy recently?

**Comment by DanisNone** [comment]
@MythicArrow  It seems that `prelu` is equivalent to `leaky_relu`?

**Comment by MythicArrow** [comment]
Yeah it looks similar but there is a significant difference between them. Leaky ReLu contains "a" as a fixed constant but when it comes to PReLu its "a" is a learnable slope parameter, which is trained through backpropagation.

**Comment by DanisNone** [comment]
You can pass a JAX array as the second argument to `leaky_relu`, and JAX will have no issues computing gradients through it.

**Comment by jakevdp** [comment]
Yeah indeed @DanisNone – it looks like this is the same as `leaky_relu`. Sorry @MythicArrow, I should have noticed that earlier.

**Comment by jakevdp** [comment]
I think given that, this PR can probably be closed.

**Comment by MythicArrow** [comment]
But the PReLu's "a" is learnable not a constant.

**Comment by MythicArrow** [comment]
I can make it eligible for being trained through backpropagation.

**Comment by jakevdp** [comment]
> But the PReLu's "a" is learnable not a constant.

That is true of `leaky_relu` as well.

**Comment by MythicArrow** [comment]
> > But the PReLu's "a" is learnable not a constant.
> 
> That is true of `leaky_relu` as well.

Oh ok

**Comment by MythicArrow** [comment]
> Yeah indeed @DanisNone – it looks like this is the same as `leaky_relu`. Sorry @MythicArrow, I should have noticed that earlier.

No problem sir.

**Comment by MythicArrow** [comment]
Ok then I will close this pr.

**Comment by MythicArrow** [comment]
I checked the LeakyReLu's proportions and I see that it has a fixed value of "a" differing from the learnable one in the PReLu so would you merge the pr if I had edited the implementation in a way that it would have made the "a" learnable and trainable through backpropagation instead of the fixed one in LeakyReLu?

**Comment by jakevdp** [comment]
You mention `LeakyReLu` – are you talking about the function in the `stax` example library? https://github.com/jax-ml/jax/blob/9678a764e83b5fab3a877348c6ccc503d9808145/jax/example_libraries/stax.py#L158
If so, then we likely wouldn't accept such a contribution, because we are not adding new features to stax.

If you're thinking of `jax.nn.leaky_relu`, then I can't say I understand your request here. The `negative_slope` is a parameter just like any other, and you can take the gradient with respect to it, which means its value can be optimized given an appropriate loss function. That makes me think it qualifies as a "learnable parameter" – is there something more you're looking for?

**Comment by MythicArrow** [comment]
I had checked the original paper of LeakyReLu and its "a" was a fixed constant like 0.01 and wasn't learned. But in this case the jax's LeakyReLu func is like PReLu so I think there is no need for implementing PReLu.

**Comment by MythicArrow** [comment]
Sir, I have researched the PReLu and found out that it indeed has a learnable parameter which is included in the model's weights. So if I define it as a parameter then it will be automatically learnable differing from the LeakyReLu. Does JAX allow to define 'a' as a model parameter?
